include/nuttx/hwspinlock/hwspinlock.h

```diff
@@ -0,0 +1,126 @@
+/****************************************************************************
+ * include/nuttx/hwspinlock/hwspinlock.h
+ *
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_NUTTX_HWSPINLOCK_HWSPINLOCK_H
+#define __INCLUDE_NUTTX_HWSPINLOCK_HWSPINLOCK_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/compiler.h>
+#include <nuttx/spinlock.h>
+
+#include <stdbool.h>
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+struct hwspinlock_dev_s;
+
+struct hwspinlock_ops_s
+{
+  CODE bool (*trylock)(FAR struct hwspinlock_dev_s *dev,
+                       int id, int priority);
+  CODE void (*relax)(FAR struct hwspinlock_dev_s *dev,
+                     int id, int priority);
+  CODE void (*unlock)(FAR struct hwspinlock_dev_s *dev, int id);
+};
+
+struct hwspinlock_dev_s
+{
+  spinlock_t lock;
+  FAR const struct hwspinlock_ops_s *ops;
+};
+
+/****************************************************************************
+ * Public Function Prototypes
+ ****************************************************************************/
+
+#ifdef __cplusplus
+#define EXTERN extern "C"
+extern "C"
+{
+#else
+#define EXTERN extern
+#endif
+
+static inline bool hwspin_trylock(FAR struct hwspinlock_dev_s *dev,
+                                  int id, int priority)
+{
+  return dev->ops->trylock(dev, id, priority);
```

Comment:
Why we can't call it directly?

Reply:
simple and unify the usage(e.g. hwspin_trylock v..s. hwspin_trylock_irqsave)

Reply:
I see. That is a separate API. My initial feeling was that it is only used below, but it is exposed to the user.